haystack_experimental/components/agents/agent.py

```
@@ -273,13 +214,38 @@ def run(
+        # Repeat until the exit condition is met
+        counter = 0
+        while counter < self.max_runs_per_component:
+            # 1. Call the ChatGenerator
+            llm_messages = self.chat_generator.run(messages=messages, **generator_inputs)["replies"]
+
+            # TODO Possible for LLM to return multiple messages (e.g. multiple tool calls)
+            #      Would a better check be to see if any of the messages contain a tool call?
+            # 2. Check if the LLM response contains a tool call
+            if llm_messages[0].tool_call is None:
+                return {"messages": messages + llm_messages, **state.data}
+
+            # 3. Call the ToolInvoker
+            # We only send the messages from the LLM to the tool invoker
+            tool_invoker_result = self._tool_invoker.run(messages=llm_messages, state=state)
+            tool_messages = tool_invoker_result["messages"]
+            state = tool_invoker_result["state"]
+
+            # 4. Check the LLM and Tool response for the exit condition, if exit_condition is a tool name
+            if self.exit_condition != "text":
+                # TODO Possible for LLM to return multiple messages (e.g. multiple tool calls)
+                #      So exit condition could be missed if it's not the first message
+                should_exit = (
+                    llm_messages[0].tool_call.tool_name == self.exit_condition
+                    and not tool_messages[0].tool_call_result.error
+                )
+                if should_exit:
+                    return {"messages": messages + llm_messages + tool_messages, **state.data}
+
+            # 5. Combine messages, llm_messages and tool_messages and send to the ChatGenerator
+            messages = messages + llm_messages + tool_messages
+            counter += 1
+
+        logger.warning(f"Agent exceeded maximum runs per component ({self.max_runs_per_component}), stopping.")
```

Question: why not raising an exception here (after exceeding `self.max_runs_per_component`)?

Maybe it would be nice to add a test for multiple subsequent tool invocation failures.

---

Good question, unsure if we should raise an exception or if we exit with a warning message. 

For example, I could see a user just want to continue the generation by calling the agent again with the output returned here. And just because we hit the max runs per component doesn't mean it's an error per say.

---

> Maybe it would be nice to add a test for multiple subsequent tool invocation failures.

I totally agree, if we are all in support of going in this direction, I'll spend the time to expand the tests :)

---

@sjrl 

> For example, I could see a user just want to continue the generation by calling the agent again with the output returned here. And just because we hit the max runs per component doesn't mean it's an error per say.

Makes sense! What do you think about adding an additional option for raising here then?

---

Yeah I think that could make sense. To clarify, we think this could be helpful for debugging purposes? So if I have an Agent in a Pipeline that we stop the execution in case this happens instead of continuing?

---

yes, IMHO it could be useful for tool debugging purposes. Example: you have a tool which is not giving a good-enough answer to a LLM or so and it will be called repeatedly (and you don't want that). But ofc one could also simply gracefully answer after N tool calls failures. Your choice, nothing strictly needed ;)